advancedlog/json_logger.go

```diff
@@ -0,0 +1,57 @@
+package logger
+
+import (
+	"encoding/json"
+	"io"
+	"time"
+)
+
+// RFC3339Micro ...
+const RFC3339Micro = "2006-01-02T15:04:05.999999Z07:0"
+
+func defaultTimeProvider() time.Time {
+	return time.Now()
+}
+
+type jsonLogger struct {
+	debugLogEnabled bool
+	encoder         *json.Encoder
+	timeProvider    func() time.Time
+}
+
+func newJSONLogger(output io.Writer, provider func() time.Time) SimplifiedLogger {
+	logger := jsonLogger{
+		debugLogEnabled: false,
+		encoder:         json.NewEncoder(output),
+		timeProvider:    provider,
+	}
+
+	return &logger
+}
+
+func (j *jsonLogger) EnableDebugLog(enabled bool) {
+	j.debugLogEnabled = enabled
+}
+
+func (j *jsonLogger) IsDebugLogEnabled() bool {
+	return j.debugLogEnabled
+}
+
+func (j *jsonLogger) LogMessage(producer Producer, level Level, message string) {
+	if j.debugLogEnabled == false && level == DebugLevel {
+		return
+	}
+
+	logMessage := logMessage{
+		Timestamp:   j.timeProvider().Format(RFC3339Micro),
+		MessageType: "log",
+		Producer:    string(producer),
+		Level:       string(level),
+		Message:     message,
+	}
+
+	err := j.encoder.Encode(logMessage)
+	if err != nil {
+		// This is only to satisfy errcheck.
```

Comment:
Shouldn't we introduce better error handling for this, like printing the error to the build log using a hand-crafted json?

Reply:
That is a good point. I was thinking that none of the parameters are complex in the sense that they are built in types  and we control the message struct so this should never fail in theory. 

But I like the idea to construct a json string manually and print it out.